tests/queries/0_stateless/03271_date_to_datetime_saturation.reference

```diff
@@ -0,0 +1,3 @@
+200
+200
+100
```

Comment:
This is not ok. It should never return incorrect values.

A couple of requirements:
* This should have way more tests than it has, explicitly testing parts with ranges (before supported Datetime, supported Datetime, after supported Datetime) and its combinations. Also with extreme values in both dates and datetime.
* It's not ok to ever return bad values in index analysis. There are 3 possible alternatives:
  - Support everything (might need a type that's a superset of both Date / DateTime for it to be possible).
  - Throw on invalid values.
  - Ignore index analysis in the part where the cast is lossy, so the part is read anyways.
The only reason to support `ignore` or `saturate` if it leads to incorrect analysis would be to support old queries that used to work incorrectly and where throwing breaks compatibility, but it should not be the default value.

All this is only possible to verify with extensive tests, so that should be introduced with the PR, not as an after thought.

Reply:
> It should never return incorrect values.

What do you mean by incorrect? The implementation of the overflow behavior or the fact that it overflows? 

And which test is not ok? This one alone
```
select count() from test where toDateTime(stamp) >= parseDateTimeBestEffort('2024-11-01')
settings date_time_overflow_behavior = 'ignore';  -- returns 100
```

or the entire test suite?

Reply:
I was talking about only this test with `ignore` but I didn't notice that you had changed the original test and now it's doing DateTime >= DateTime instead of the original Date > DateTime.

My main point is that implicit conversion for index analysis should never lead to skip valid parts. So either the conversion is lossless or, for corner cases that can't be covered, it reads the whole part (maybe with a profile event to count these currently impossible analysis). If that's how it will always work, great.

Let's please add tests with different Date vs DateTime comparison, with and without index analysis (`ORDER BY tuple()` vs `ORDER BY col`), and with lossy vs lossless conversion, extreme values and so on. That way we ensure the results are correct and we can iterate in the future if you decide to improve it somehow with some common supertype.

Reply:
> implicit conversion for index analysis should never lead to skip valid parts.

The purpose of this PR is to ensure this exactly. And the corner cases are covered in `isFunctionReallyMonotonic`.

> Let's please add tests with different Date vs DateTime comparison, with and without index analysis 

Sure. BTW, there's no need to use `ORDER BY tuple()` here. Traditionally, the `identity` function is used for this purpose.